FORMS.md

```diff
@@ -0,0 +1,243 @@
+
+- `active`: When the Input is activated by the user.
+- `autofill`: When an Input has been auto-filled by the browser.
+- `blank`: When an Input is empty.
+- `checked`: When Inputs like Radio and Checkbox are checked.
+- `dir`: The reading direction currently applied.
+- `disabled`: When an Input is disabled.
+- `enabled`: When an Input is enabled (rather than disabled). This is the default state.
+- `focus-within` When a child has `focus`.
+- `focus`: When the Input has focus.
+- `hover`: When the Input is hovered by a pointing device.
+- `in-range` When an Input is within the range constrictions. e.g. a slider.
+- `indeterminate`: When Inputs are in an indeterminate state. e.g. A "parent" checkbox for a group of mixed checked/unchecked Inputs.
+- `invalid`: When an Input has invalid contents/values.
+- `optional`: When an Input is optional.
+- `out-of-range`: When an Input is outside of range constrictions.
+- `placeholder-shown`: When placeholder text is visible.
+- `read-only`: When an Input cannot be changed by the user.
+- `read-write`: When an Input is user-editable (rather than read-only). This is the default state.
+- `required`: When an Input is required.
+- `target-within`: When a child is a `target`.
+- `target`: When the component is the target of a URL. e.g. `site.com/form#firstname`
+- `user-invalid`: When an Input is invalid, but only after a user has interacted with it.
+- `valid`: When an Input has valid contents/values.
+
+### Content
+
+- FieldLabel: a text label for the field (element may be `<label>` or `<legend>` depending on Field input)
+  - The Label also indicates the Field's optional/required status
+- FieldHint: context for the field
+- Input: the UI control
+- FieldValidation: any error or success state messaging for the field
+
+### Psuedo-structure
+
+```html
+// via props
+
+<Field
+  label="A field"
+  hint="This is a field you should fill in."
+  type="text"
+  name="a-field"
+  optional
+  disabled
+  autocomplete="fname"
+/>
+
+
+// via composition
+
+<Field>
+  <Field.Label>A field</Field.Label>
+  <Field.Hint>This is a field you should fill in.</Field.Hint>
+  <Input/>
+  <Field.Validation type="error">This is an error message.</Field.Validation>
+</Field>
+```
+
+### HTML
+
+```html
+<fieldset class="ods-field">
+  <label class="ods-field--label" for="first-name">
+    First name
+    <span class="ods-field--optional">Optional</span>
+  </label>
+  <aside class="ods-field--hint" id="first-name-hint">
+    Please enter your first name.
+  </aside>
+  <input class="ods-text-input" type="text" name="first-name" id="first-name" spellcheck="false" value="" autocomplete="fname" aria-describedby="first-name-hint first-name-error">
+  <aside class="ods-field--error" id="first-name-error"><span class="ods-u-visually-hidden">Error:</span> Numbers are disallowed.</aside>
+</fieldset>
```

Comment:
This new structure definitely addresses some of the issues we had before! My questions here are mainly around the use of the `aside` and `fieldset` tags. 

[<aside>: The Aside element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/aside)
> The <aside> HTML element represents a portion of a document whose content is only indirectly related to the document's main content. Asides are frequently presented as sidebars or call-out boxes.

[<fieldset>: The Field Set element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/aside)
> The <fieldset> HTML element is used to group several controls as well as labels (<label>) within a web form.

I'm not sure we meet the criteria for either one of these given the use case above. Fieldset is particularly interesting because my gut tells me that it's meant for grouping a collection of elements like checkbox and radio groups, not a single input element.

Definitely want to expand on these questions, this is certainly an accessibility concern!

Reply:
After looking at `aside`'s implicit ARIA role (<a href="https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Complementary_role">complimentary</a>), I'm wholly agreed. The content definitely can't stand on it's own and should be considered part of the main flow. `p` might be sufficient for these two use cases, especially since the content shouldn't extend beyond a single paragraph.

`fieldset`'s implicit `group` role makes sense to me here, <a href="https://w3c.github.io/aria/#group">given the definition</a>. You're right about the intended purpose, though. It's a shame that HTML doesn't offer anything akin to `field`. I'd be comfortable using `div` here since no a11y concerns would be affected. That would also answer the question of where shared attributes go: directly on the input.

Worth noting, `fieldset` is required for inputs like RadioButton, but we could have conditional HTML for associated input types the way we do with `label` and `legend`.

Reply:
I always thought of fieldset as appropriate for checkboxes and radio groups, or in [this instance](https://www.w3.org/WAI/tutorials/forms/grouping/#related-fields), where you use a legend to describe a collection of inputs, the example uses Billing and Shipping Address.

If we use in fieldset in the way described in this doc, how does that affect the Billing/Shipping address use case?

Reply:
They can be nested, so I planned on using them for both `Field` and `FieldGroup`, which would be used for the use case above.

Since React can render the correct elements dynamically, I can update this to include notes for single- and multi-part Inputs. That is, for Checkbox/Radio, output `fieldset` as the `Field` wrapper; for anything else, `div`.

Reply:
+1 `<fieldset>` for multiple input groupings whether that be a associated radio inputs or separate text inputs each with their own name. For grouping the contents of a single field consider using `label` as the root and my comment above related to that